### PR TITLE
add removeshorts arg and supply cookie manualy

### PR DESCRIPTION
**Pull request by indrastorms**
Title explains it

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -75,6 +75,20 @@ def main():
...
     )
 
+    parser.add_argument(
+        "--cookiefile",
+        type=str,
+        default=None,
+        help="Path of netscaped cookie file.",
+    )
+
+    parser.add_argument(
```
Comment (oSumAtrIX) [inline main.py]: This shouldnt be an option. Shorts dont need to be considered.

File: main.py [hunk 2]
```diff
@@ -75,6 +75,20 @@ def main():
         help="Browser to use for cookies.",
     )
 
+    parser.add_argument(
```
Comment (oSumAtrIX) [inline main.py]: What if both the browser and cookie file are specified?

Comment (indrastorms) [reply main.py]: It tries to open it in browser and get failed

File: main.py (hunk 1 above)
Comment (indrastorms) [reply main.py]: There is feat request so added it as a bonus

File: main.py (hunk 2 above)
Comment (oSumAtrIX) [reply main.py]: Who is it? What does it open? And what gets failed?

File: main.py (hunk 1 above)
Comment (oSumAtrIX) [reply main.py]: Not needed, you can remove the option and just skip shorts to make the script faster

File: main.py (hunk 2 above)
Comment (indrastorms) [reply main.py]: Who: Ytdlp
How: see the crashlog reported

File: main.py [hunk 3]
```diff
@@ -92,6 +108,15 @@ def main():
...
 
+    if REMOVE_SHORTS is True:
+        popped = 0
+        for index in reversed(range(len(data))):
+            if data[index]["title"].lower().find("short") > 0:
+                data.pop(index)
+                print(f"popped {data[index]['title']}")
+                popped += 1
+        print(f"total popped {popped}")
+
```
Comment (guillaumematheron) [inline main.py]: Suggestions:
 - Better wording for the prints
 - `if REMOVE_SHORTS is True` -> `if REMOVE_SHORTS`
 - I'm not sure why this needs to be handled here and not in `filter_video_events` ? This looks more readable to me:
```
if '#shorts' in event['title']:
    continue
```

File: main.py (hunk 2 above)
Comment (oSumAtrIX) [reply main.py]: The crashlog was not caused by the cookie file as no cookie file was provided. My question is 

> What if both the browser and cookie file are specified?

File: main.py [hunk 4]
```diff
@@ -170,9 +196,15 @@ def download_videos(
...
         "cookiesfrombrowser": (COOKIES_FROM_BROWSER,),
+        "cookiefile": COOKIE_FILE,
         "format": "worstaudio"
     }
 
+    if COOKIE_FILE != None:
+        opts.pop("cookiesfrombrowser")
+    else:
+        opts.pop("cookiefile")
+
```
Comment (guillaumematheron) [inline main.py:192]: Not sure what this does?

File: main.py (hunk 1 above)
Comment (indrastorms) [reply main.py]: Default is false so, it doesn't effect speed of execution unless explicitly said.
But still if you hate it we can remove it

File: main.py (hunk 4 above)
Comment (indrastorms) [reply main.py:192]: It removes the key from opts dict

File: main.py (hunk 1 above)
Comment (oSumAtrIX) [reply main.py]: No I am not saying that speed will differ, I am saying to skip shorts, the option itself has no usecase, there's no scenario where one would want to watch shorts

File: main.py (hunk 4 above)
Comment (guillaumematheron) [reply main.py:192]: Ok but why ?

Comment (guillaumematheron) [reply main.py:192]: Ah to make sure they are exclusive...

In that case I suggest https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group so the code and `--help` are more readable

File: main.py (hunk 2 above)
Comment (indrastorms) [reply main.py]: When we giving cookiesfrombrowser arg ytdlp tries to handle it no matter cookiefile is supplied or not. And in termux the cookiesfrombrowser doesn't get things and get crashed.

File: main.py (hunk 4 above)
Comment (indrastorms) [reply main.py:192]: But it doesn't help to remove the if block right?

File: main.py (hunk 3 above)
Comment (indrastorms) [reply main.py]: Removed entirely

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
I think this could work. Whats missing is adjusting the README so that it mentions that a cookiefile can be used and how.

**Comment by indrastorms** [comment]
> I think this could work. Whats missing is adjusting the README so that it mentions that a cookiefile can be used and how.

Can't help, even I can't understand my sentence sometimes 😅

Btw the addon name is "cookie editor" but I think you need firefox nightly or kiwi(not tested)

**Comment by indrastorms** [comment]
@guillaumematheron one question does this script download the whole video to write the history?

**Comment by oSumAtrIX** [comment]
Yes. https://github.com/ReVanced/restore-missing-youtube-watch-history downloads the whole lowest audio. I think you can also just download the m3u8 instead even faster.

**Comment by indrastorms** [comment]
> Yes. https://github.com/ReVanced/restore-missing-youtube-watch-history downloads the whole lowest audio. I think you can also just download the m3u8 instead even faster.

That explains why too much data is used. Can't we stop the download and continue to next video when certain amount of bytes is downloaded to write the history.

**Comment by guillaumematheron** [comment]
I'm not quite sure, I used the `simulate` argument https://github.com/ReVanced/restore-missing-youtube-watch-history/blob/main/main.py#L170 but the man page doesn't explicitly say how this is achieved https://man.archlinux.org/man/yt-dlp.1#s,

**Comment by oSumAtrIX** [comment]
@Indranil012 Asking YTDL to download the m3u8 should be enough. Downloading the full track is necessary to prevent the video from appearing on YouTube.

**Comment by guillaumematheron** [comment]
I think that's the implementation https://github.com/yt-dlp/yt-dlp/blob/cf91400a1dd6cc99b11a6d163e1af73b64d618c9/yt_dlp/extractor/youtube.py#L3197

**Comment by oSumAtrIX** [comment]
@alexandreteles Is this something that can be called?

**Comment by alexandreteles** [comment]
> @alexandreteles Is this something that can be called?

It could be called if we *really* wanted it but it is clearly marked as an internal method. Ignoring the internal method issue for a moment, we could do something like this considering the `YoutubeIE` class:

```py
from urllib.parse import urlparse
from yt_dlp.utils import unsmuggle_url
from yt_dlp.extractor.youtube import YoutubeIE

youtube = YoutubeIE()

initial_url = "https://www.youtube.com/watch?v=EgRqkOLxEOE"

url, smuggled_data = unsmuggle_url(initial_url, {})

video_id = youtube._match_id(url)

base_url = f"{urlparse(url).scheme}://www.youtube.com/"

webpage_url = f"{base_url}watch?v={video_id}"

_, _, player_responses, _ = youtube._download_player_responses(
	url, smuggled_data, video_id, webpage_url
)

_mark_watched(video_id, player_responses)
```

@oSumAtrIX would you or someone else be able to test it?

**Comment by indrastorms** [comment]
We can supply "mark_watched" = True through opts.
https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/options.py#L438

**Comment by alexandreteles** [comment]
> We can supply "mark_watched" = True through opts. [yt-dlp/yt-dlp@`master`/yt_dlp/options.py#L438](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/options.py?rgh-link-date=2024-03-06T02%3A41%3A36Z#L438)

I just tried:

```py
import yt_dlp

url: str = "https://www.youtube.com/watch?v=EgRqkOLxEOE"

opts: dict[ str, str | set[str] ] = {
	"simulate": True,
	"mark_watched": True,
	"format": "worstaudio",
	"cookiesfrombrowser": ("chrome",)
}

with yt_dlp.YoutubeDL(ytdl_opts) as runner:
	runner.download(url)
```

And it works. You might need to add to the documentation that if you are extracting from Edge you might need to run as admin or you will get a permission error on the cookies file. The `download` method accepts an iterable of strings but they are processed synchronously so we might need to write an async loop ourselves.

**Comment by oSumAtrIX** [comment]
@alexandreteles the permission error is caused by a running process. The steps can include a step to kill the browser. Off topic for this PR though

**Comment by alexandreteles** [comment]
Maybe something like this could work:

```py
import yt_dlp
import asyncio
from typing import Any
from concurrent.futures import ThreadPoolExecutor


async def worker(string: str, semaphore: asyncio.Semaphore, work_function: callable, loop: asyncio.AbstractEventLoop) -> Any:
    async with semaphore:
        result = await loop.run_in_executor(None, lambda: work_function(string))
        return result


async def main():
	# Allow max 3 concurrent executions
    semaphore = asyncio.Semaphore(3)

    # the set of URLs we got from the JSON file
    urls: set[str] = ("url1", "url2", "url3", "url4", "url5", "url6")

    opts: dict[str, str | set[str]] = {
        "simulate": True,
        "mark_watched": True,
        "format": "worstaudio",
        "cookiesfrombrowser": ("chrome",),
    }

    ydl = yt_dlp.YoutubeDL(opts)

    with ThreadPoolExecutor() as _:
        loop = asyncio.get_running_loop()
        tasks = map(lambda url: worker(url, semaphore, ydl.download, loop), urls)
        results = await asyncio.gather(*tasks)
        print(results)


if __name__ == "__main__":
    asyncio.run(main())
```

**Review by oSumAtrIX — APPROVED** [review]

File: main.py [hunk 5]
```diff
@@ -108,15 +102,6 @@ def main():
     with WATCH_HISTORY_FILE.open(encoding="utf8") as f:
         data = json.load(f)
 
-    if REMOVE_SHORTS is True:
```
Comment (oSumAtrIX) [inline main.py]: @Indranil012 Can you open a separate PR to filter shorts? These are not necessary.

Comment (indrastorms) [reply main.py]: Ok

**Pull request merged**